File: 2022/day12/day12.rb
#!/usr/bin/env ruby
require 'set'

map = ARGF.map { |line| line.chars }

def elevation(letter)
    letter.tr('SE','az').ord - 'a'.ord
end

alternates = []
edges = {}
start = nil
dest = nil

map.each_with_index do |row, i|
  row.each_with_index do |letter, j|
    coords = "#{i},#{j}"
    if letter == 'S' then
      start = coords
    elsif letter == 'E' then
      dest = coords
    elsif letter == 'a' then
      alternates << coords
    end
    level = elevation(letter)
    [ [-1,0], [0,1], [1,0], [0,-1] ].each do |di, dj|
      ni = i + di
      if 0 <= ni and ni < map.length then
        nj = j + dj
        if 0 <= nj and nj < map[ni].length then
          nlevel = elevation(map[ni][nj])
          edges[coords] ||= []
          edges[coords] << "#{ni},#{nj}" if nlevel <= level + 1
        end
      end
    end
  end
end

2.times do |part|
  heads = Set.new([start])
  heads += alternates if part > 0

  distance = 0
  while !heads.include? dest do
    distance += 1
    heads.entries.each do |node|
      edges[node].each do |neighbor|
        heads << neighbor
      end
    end
  end

  puts "Part #{part+1}: #{distance}";
end

File: 2022/day12/day12.js
#!/usr/bin/env node
const fs = require('fs')

const map = fs.readFileSync(process.argv[2]).toString().split("\n").map(s => s.split(''));

const elevation = letter => {
  if (letter == 'S') {
    letter = 'a'
  } else if (letter == 'E') {
    letter = 'z'
  }
  return letter.charCodeAt(0) - 'a'.charCodeAt(0)
}

let alternates = []
let edges = {}
let start, end

for (const [i, row] of map.entries()) {
  for (const [j, letter] of row.entries()) {
    const coords = `${i},${j}`
    if (letter == 'S') {
      start = coords
    } else if (letter == 'E') {
      end = coords
    } else if (letter == 'a') {
       alternates.push(coords)
    }

    const level = elevation(letter)
    for (const [di, dj] of [[-1,0],[0,1],[1,0],[0,-1]]) {
      const ni = i + di
      if (0 <= ni  && ni < map.length) {
        const nj = j + dj
        if (0 <= nj && nj < map[ni].length) {
          const nlevel = elevation(map[ni][nj])
          if (nlevel <= level + 1) {
            (edges[coords]||=[]).push(`${ni},${nj}`)
          }
        }
      }
    }
  }
}

for (const part of [0,1]) {
  let heads = new Set([start])
  if (part)
    heads = new Set([...heads, ...alternates])

  let distance = 0;
  while (!heads.has(end)) {
    distance += 1
    for (const node of [...heads]) {
      for (const neighbor of (edges[node]||[])) {
        heads.add(neighbor)
      }
    }
  }

  console.log(`Part ${part+1}: ${distance}`)
}
